### ISSUE TITLE
wow, what a clever implementation of the forward algorithm on normal-space instead of log-space

### ISSUE THREAD

**Issue by sustcsonglin** [justinchiu/low-rank-models#1]
enjoy reading it

**Comment by justinchiu** [comment]
Thanks! Let us know if you have any questions. Ideally we could perform the forward algorithm in log-space for numerical stability as the number of states increases, but gemm is just too fast.

**Comment by sustcsonglin** [comment]
Yeah, logsumexp is slow and GPU memory-consuming.  I tried to use genbmm to cope with the memory issues but found it slower, lol

BTW, I wonder that in your current implementation, what is the upper bound of the number of states if there is no numerical instability issue?  I think you do a great job to solve the numerical issue by normalizing the forward score at each time step, but I do not know if it could eliminate the numerical issue at all or to some extent.

**Comment by justinchiu** [comment]
Regarding numerical instability: Great question. Our experiments with 32k state HMMs only obtained the same performance as 16k states, which I hypothesize is due to numerical stability.

The reason I believe this is that a 32k state sparse emission HMM (from prior work), which only needs to normalize over K = 256 << 32k states per timestep, outperforms a 16k state unconstrained HMM (using the renormalization trick in the standard semiring). However this is not exactly a direct comparison, as emission sparsity could serve to bias the model as well.

**Comment by sustcsonglin** [comment]
Thank you for replying~ It is very helpful
